src/NuGet.Services.AzureSearch/VersionList/MutableHijackIndexDocument.cs

```diff
@@ -0,0 +1,162 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+
+namespace NuGet.Services.AzureSearch
+{
+    /// <summary>
+    /// A mutable version of <see cref="HijackIndexDocument"/>. The booleans in this class are nullable so that we can
+    /// track the "undetermined" state. Once the booleans are set to true or false, they cannot be changed again. This
+    /// helps protect against bugs in the code that calls
+    /// <see cref="ApplyChange(SearchFilters, HijackIndexChangeType)"/> in an inconsistent manner. Technically, the
+    /// <see cref="Delete"/> and <see cref="UpdateMetadata"/> do not need to be nullable because there is no code path
+    /// that can set them explicitly to false. However it's better to be consistent with the latest booleans and employ
+    /// the same strategy. Null booleans can be assumed by the caller to be false.
+    /// </summary>
+    internal class MutableHijackIndexDocument : IEquatable<MutableHijackIndexDocument>
+    {
+        public MutableHijackIndexDocument()
+        {
+        }
+
+        public MutableHijackIndexDocument(
+            bool? delete,
+            bool? updateMetadata,
+            bool? latestStableSemVer1,
+            bool? latestSemVer1,
+            bool? latestStableSemVer2,
+            bool? latestSemVer2)
+        {
+            Delete = delete;
+            UpdateMetadata = updateMetadata;
+            LatestStableSemVer1 = latestStableSemVer1;
+            LatestSemVer1 = latestSemVer1;
+            LatestStableSemVer2 = latestStableSemVer2;
+            LatestSemVer2 = latestSemVer2;
+        }
+
+        public bool? Delete { get; private set; }
+        public bool? UpdateMetadata { get; private set; }
+        public bool? LatestStableSemVer1 { get; private set; }
+        public bool? LatestSemVer1 { get; private set; }
+        public bool? LatestStableSemVer2 { get; private set; }
+        public bool? LatestSemVer2 { get; private set; }
+
+        public void ApplyChange(SearchFilters searchFilters, HijackIndexChangeType changeType)
+        {
+            bool latest;
+            switch (changeType)
+            {
+                case HijackIndexChangeType.Delete:
+                    Guard.Assert(
+                        !Delete.HasValue || Delete.Value,
```

Comment:
`Delete != false` might be a bit more readable.

Reply:
Oh, nice point.